planning/autoware_manual_lane_change_handler/src/manual_lane_change_handler.cpp

```diff
@@ -0,0 +1,248 @@
+// limitations under the License.
+
+#include "manual_lane_change_handler.hpp"
+
+#include <deque>
+#include <string>
+#include <vector>
+
+namespace autoware::manual_lane_change_handler
+{
+
+std::vector<autoware_planning_msgs::msg::LaneletPrimitive>
+ManualLaneChangeHandler::sortPrimitivesLeftToRight(
+  const route_handler::RouteHandler & route_handler,
+  autoware_planning_msgs::msg::LaneletPrimitive preferred_primitive,
+  std::vector<autoware_planning_msgs::msg::LaneletPrimitive> primitives)
+{
+  using Primitive = autoware_planning_msgs::msg::LaneletPrimitive;
+
+  std::deque<Primitive> sorted_primitives;
+
+  auto find_primitive = [&](lanelet::Id id) -> std::optional<Primitive> {
+    auto it = std::find_if(
+      primitives.begin(), primitives.end(), [&](const Primitive & p) { return p.id == id; });
+    if (it != primitives.end()) return *it;
+    return std::nullopt;
+  };
+
+  lanelet::ConstLanelet current = route_handler.getLaneletsFromId(preferred_primitive.id);
+  // Walk left lanes
+  for (auto left = route_handler.getLeftLanelet(current, true); left;
+       left = route_handler.getLeftLanelet(*left, true)) {
+    if (auto match = find_primitive(left->id())) {
+      sorted_primitives.push_front(*match);
+    }
+  }
+
+  sorted_primitives.push_back(preferred_primitive);
+
+  // Walk right lanes
+  for (auto right = route_handler.getRightLanelet(current, true); right;
+       right = route_handler.getRightLanelet(*right, true, true)) {
+    if (auto match = find_primitive(right->id())) {
+      sorted_primitives.push_back(*match);
+    }
+  }
+
+  std::vector<autoware_planning_msgs::msg::LaneletPrimitive> result{
+    sorted_primitives.begin(), sorted_primitives.end()};
+
+  return result;
+}
+
+LaneChangeRequestResult ManualLaneChangeHandler::process_lane_change_request(
+  const int64_t ego_lanelet_id, const SetPreferredLane::Request::SharedPtr req)
+{
+  const DIRECTION override_direction =
+    req->lane_change_direction == SetPreferredLane::Request::LEFT    ? DIRECTION::MANUAL_LEFT
+    : req->lane_change_direction == SetPreferredLane::Request::RIGHT ? DIRECTION::MANUAL_RIGHT
+                                                                     : DIRECTION::AUTO;
+  if (override_direction == DIRECTION::AUTO) {
+    // Use back-up
+    if (!original_route_) {
+      return {
+        {},
+        false,
+        "Manual lane selection to AUTO is commanded but canceled due to no original route "
+        "available."};
+    }
+
+    std::vector<LaneletPrimitive> preferred_primitives;
+    std::transform(
+      original_route_.value()->segments.begin(), original_route_.value()->segments.end(),
+      std::back_inserter(preferred_primitives),
+      [](const auto & segment) { return segment.preferred_primitive; });
+    original_route_ = std::nullopt;
+
+    return {
+      preferred_primitives, true,
+      "Manual lane selection to AUTO is commanded and executed successfully."};
```

Comment:
I rather would like to have this original route handling to be done in the mission_planner side rather than in manual lane change handler.

Manual lane change handler would set "reset" field to to true in the service and send that to mission_planner. mission_planner will reset back to original route whenever it receives the reset request.